### ISSUE TITLE
NullPointerException LockResourceImpl.java:166  (3.8-SANPSHOT)

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#9096]
I have encountered this NPE

```
java.lang.NullPointerException
        at com.hazelcast.concurrent.lock.LockResourceImpl.canAcquireLock(LockResourceImpl.java:166)
        at com.hazelcast.concurrent.lock.LockStoreImpl.canAcquireLock(LockStoreImpl.java:159)
        at com.hazelcast.concurrent.lock.operations.LockOperation.shouldWait(LockOperation.java:69)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.waitingNeeded(OperationRunnerImpl.java:235)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:177)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:373)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:400)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:283)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:268)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:232)
        at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:207)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:59)
        at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:64)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:137)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:127)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102)
        at ------ submitted from ------.(Unknown Source)
        at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:114)
        at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:241)
        at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.execute(AbstractPartitionMessageTask.java:78)
        at com.hazelcast.spi.impl.AbstractInvocationFuture.unblock(AbstractInvocationFuture.java:237)
        at com.hazelcast.spi.impl.AbstractInvocationFuture.andThen(AbstractInvocationFuture.java:213)
        at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:69)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
        at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
        at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:137)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:127)
        at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:102)
        at ------ submitted from ------.(Unknown Source)
        at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrow(ClientInvocationFuture.java:74)
        at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrow(ClientInvocationFuture.java:30)
        at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:158)
        at com.hazelcast.client.spi.ClientProxy.invokeOnPartition(ClientProxy.java:153)
        at com.hazelcast.client.proxy.PartitionSpecificClientProxy.invokeOnPartition(PartitionSpecificClientProxy.java:47)
        at com.hazelcast.client.proxy.ClientLockProxy.tryLock(ClientLockProxy.java:134)
        at com.hazelcast.client.proxy.ClientLockProxy.tryLock(ClientLockProxy.java:124)
        at hz.lock.CreateUseDestroy.timeStep(CreateUseDestroy.java:18)
        at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:61)
        at remote.bench.marker.MetricsMarker.timeBenchFlatOut(MetricsMarker.java:45)
        at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:36)
        at remote.bench.BenchThread.call(BenchThread.java:38)
        at remote.bench.BenchThread.call(BenchThread.java:12)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
```

jenkins build job https://hazelcast-l337.ci.cloudbees.com/view/multi-version/job/createUseDestroy/

this exception seems to be provoked by
HazelcastClient 3.7
HazelcastClient 3.7.1
connected to a 3-8-SANPSHOT member.

I have not seen the issue provoked by a  3-8-SANPSHOT client.
however the issue seems moderately rare to encounter.

the test simply repeatedly runs

```
        ILock lock = hzInstance.getLock(name+random.nextInt(max));
        try {
            lock.tryLock(25, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        lock.destroy();
```

from all clients.  and the test ignore's
com.hazelcast.spi.exception.DistributedObjectDestroyedException:
java.lang.IllegalStateException:
com.hazelcast.core.HazelcastInstanceNotActiveException

**Comment by Danny-Hazelcast** [comment]
could it be that 
com.hazelcast.concurrent.lock.LockResourceImpl.canAcquireLock(LockResourceImpl.java:166)
has been changed, and this issue is fixed in hat master ?

**Comment by Danny-Hazelcast** [comment]
a link to the logs config and scripts of the run http://54.87.52.100/~jenkins/workspace/createUseDestroy/3.8-SNAPSHOT/2016_10_12-10_02_40/

**Comment by Danny-Hazelcast** [comment]
Interestingly when i run this lock create use destroy test in isolation we get a difference with error

https://hazelcast-l337.ci.cloudbees.com/view/multi-version/job/create-use-destroy-lock/2/console

and the 3.8-SNAPSHOT clients quickly hits a OOME
output/HZ/HzClient1HZ/HzClient1HZ.hprof

http://54.87.52.100/~jenkins/workspace/create-use-destroy-lock/3.8-SNAPSHOT/2016_10_11-21_59_24/

http://54.87.52.100/~jenkins/workspace/create-use-destroy-lock/3.8-SNAPSHOT/2016_10_11-21_59_24/output/HZ/HzClient1HZ/
however again the hprof produced is small.


this seem similar to the issue https://github.com/hazelcast/hazelcast/issues/9062
which caused a client OOME, 
however, this issue above is using member graceful shutdown, and cluster and client all at the same version.

**Comment by jerrinot** [comment]
@Danny-Hazelcast are you sure you are using the latest master? the OOM is very suspicious as it should be fixed by https://github.com/hazelcast/hazelcast/pull/9091

**Comment by Danny-Hazelcast** [comment]
@jerrinot,  ok in fact its @ihsandemir  branch, which was based on yesterdays hat master.
but will re run with hat master.

**Comment by Danny-Hazelcast** [comment]
added client tag, as maybe its more of a client side issue.  

reproduced this issue, links

http://54.87.52.100/~jenkins/workspace/createUseDestroy/3.8-SNAPSHOT/2016_11_02-12_20_16/output/HZ/HzClient7HZ/exception.txt

https://hazelcast-l337.ci.cloudbees.com/view/multi-version/job/createUseDestroy/9/console

**Comment by jerrinot** [comment]
closed by #9566
